pennylane_lightning/lightning_qubit/_adjoint_jacobian.py

```diff
@@ -0,0 +1,344 @@
+        of a tape.
+
+        .. code-block:: python
+
+            vjp_f = dev.vjp([qml.state()], grad_vec)
+            vjp = vjp_f(tape)
+
+        computes :math:`w = (w_1,\\cdots,w_m)` where
+
+        .. math::
+
+            w_k = \\langle v| \\frac{\\partial}{\\partial \\theta_k} | \\psi_{\\pmb{\\theta}} \\rangle.
+
+        Here, :math:`m` is the total number of trainable parameters, :math:`\\pmb{\\theta}`
+        is the vector of trainable parameters and :math:`\\psi_{\\pmb{\\theta}}`
+        is the output quantum state.
+
+        Args:
+            measurements (list): List of measurement processes for vector-Jacobian product.
+                Now it must be expectation values or a quantum state.
+            grad_vec (tensor_like): Gradient-output vector. Must have shape matching the output
+                shape of the corresponding tape, i.e. number of measurements if
+                the return type is expectation or :math:`2^N` if the return type is statevector
+
+        Returns:
+            The processing function required to compute the vector-Jacobian products of a tape.
+        """
+        if tape.shots is not None:
+            warn(
+                "Requested adjoint differentiation to be computed with finite shots. "
+                "The derivative is always exact when using the adjoint differentiation "
+                "method.",
+                UserWarning,
+            )
+
+        measurements = tape.measurements
+        tape_return_type = self._check_adjdiff_supported_measurements(measurements)
+
+        if qml.math.allclose(grad_vec, 0) or tape_return_type is None:
+            return qml.math.convert_like(np.zeros(len(tape.trainable_params)), grad_vec)
+
+        if tape_return_type is Expectation:
+            if len(grad_vec) != len(measurements):
+                raise ValueError(
+                    "Number of observables in the tape must be the same as the "
+                    "length of grad_vec in the vjp method"
+                )
+
+            if np.iscomplexobj(grad_vec):
+                raise ValueError(
+                    "The vjp method only works with a real-valued grad_vec when the "
+                    "tape is returning an expectation value"
+                )
+
+            ham = qml.Hamiltonian(grad_vec, [m.obs for m in measurements])
+
+            def processing_fn_expval(tape):
+                nonlocal ham
+                num_params = len(tape.trainable_params)
+
+                if num_params == 0:
+                    return np.array([], dtype=self.qubit_state.dtype)
+
+                new_tape = qml.tape.QuantumScript(
+                    tape.operations,
+                    [qml.expval(ham)],
+                    shots=tape.shots,
+                    trainable_params=tape.trainable_params,
+                )
+
+                return self.calculate_jacobian(new_tape)
+
+            return processing_fn_expval(tape)
```

Comment:
```suggestion
            return processing_fn_expval
```

The docstring says that this generates the processing function to compute the VJP.

Reply:
The docstring is wrong. I changed that. Thank you for pointing this out.